pandas/tests/indexing/multiindex/test_multiindex.py

```diff
@@ -111,3 +111,13 @@ def test_nested_tuples_duplicates(self):
         df3 = df.copy(deep=True)
         df3.loc[[(dti[0], "a")], "c2"] = 1.0
         tm.assert_frame_equal(df3, expected)
+
+    def test_multiindex_from_product_contains_na(self):
+        # https://github.com/pandas-dev/pandas/issues/31883
+        values1 = [np.array([0.0, pd.NA], dtype="object"), ["a", "b"]]
+        values2 = [np.array([0.0, np.nan], dtype="object"), ["a", "b"]]
```

Comment:
wait, pd.NA is actually converted to np.nan here?

Reply:
Yes, probably not ideal (but better than an error). If merged would a follow-up issue to make sure pd.NA is used make sense? Or I could mark that the referenced issue is not actually closed and comment there.

Reply:
no, pls do it here

Reply:
followups are ok, but for relatively small things just fixing it in the same PR is better

Reply:
I'd have to look more closely, but I'm not sure if having it return pd.NA instead of np.nan is an easy fix; this is already how it behaves for list input (which seems to be the documented behavior):

```python
In [1]: import pandas as pd 
   ...:  
   ...: values = ["a", pd.NA] 
   ...:  
   ...: pd.Categorical(values) 
   ...:                                                                                                                            
Out[1]: 
[a, NaN]
Categories (1, object): [a]

In [2]: pd.__version__                                                                                                             
Out[2]: '1.0.1'
```

I think having it so that we at least get the same output and not an error for a numpy array with object dtype is still an improvement though? What are your thoughts @WillAyd ?

Reply:
Agree it would be nice to maintain pd.NA  - do you know the extra effort involved to do so?

Reply:
I'm not sure, I'd need to investigate a bit more. The logic doesn't seem too obvious though; should `pd.NA` be the default for missing values, or only when it's explicitly encountered? How should mixed missing value types get handled during construction (e.g., `pd.Categorical(["a", np.nan, pd.NA])`)?

Reply:
Curious if @jorisvandenbossche has a preference? Always using `pd.NA` seems logical, probably just a question of "when" to implement that (as a lot of people are likely using `Categorical` and expecting to see `NaN`).